### PR TITLE
pid1: simplify drastically how we watch bus names for service's BusName= setting

### PR DESCRIPTION
**Pull request by poettering**
Let's remove a lot of code and simplify things. Also, fix a memory leak.

(And while we are at it, also remove another synchronous bus operation, just like that)

(I did this to fix a deadlock casued by the synchronous bus operation, for #14375, but it has plenty merit on its own)

### CODE REVIEW

**Comment by mrc0mmand** [comment]
`TEST-43-PRIVATEUSER-UNPRIV_FAIL` started failing consistently with this PR:

```
Dec 23 20:54:31 systemd-testsuite testsuite.sh[298]: + runas nobody systemctl --user --wait is-system-running
Dec 23 20:54:31 systemd-testsuite testsuite.sh[298]: + declare userid=nobody
Dec 23 20:54:31 systemd-testsuite testsuite.sh[298]: + shift
Dec 23 20:54:31 systemd-testsuite testsuite.sh[298]: + su nobody -c 'XDG_RUNTIME_DIR=/run/user/$UID "$@"' -- sh systemctl --user --wait is-system-running
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.0 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=70 reply_cookie=0 signature=ss error-name=n>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.0 path=n/a interface=n/a member=n/a cookie=111 reply_cookie=70 signature=v error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.0 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=71 reply_cookie=0 signature=ss error-name=n>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.0 path=n/a interface=n/a member=n/a cookie=112 reply_cookie=71 signature=v error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.0 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=72 reply_cookie=0 signature=ss error-name=n>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.0 path=n/a interface=n/a member=n/a cookie=113 reply_cookie=72 signature=v error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.3 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=GetDynamicUsers cookie=2 reply_cookie=0 signature=n/a error-name=n/a error-messag>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.3 path=n/a interface=n/a member=n/a cookie=114 reply_cookie=2 signature=a(us) error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.0 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=73 reply_cookie=0 signature=ss error-name=n>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.0 path=n/a interface=n/a member=n/a cookie=115 reply_cookie=73 signature=v error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: user@65534.service: User lookup succeeded: uid=65534 gid=65534
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=116 reply_cookie=0 signature=sa{sv}as error-name=n/a er>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=117 reply_cookie=0 signature=sa{sv}as error-name=n/a er>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.0 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=74 reply_cookie=0 signature=ss error-name=n>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.0 path=n/a interface=n/a member=n/a cookie=118 reply_cookie=74 signature=v error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: systemd-journald.service: Got notification message from PID 262 (FDSTORE=1)
Dec 23 20:54:31 systemd-testsuite systemd[1]: systemd-journald.service: Added fd 51 (n/a) to fd store.
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.0 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=75 reply_cookie=0 signature=ss error-name=n>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.0 path=n/a interface=n/a member=n/a cookie=119 reply_cookie=75 signature=v error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.0 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=76 reply_cookie=0 signature=ss error-name=n>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.0 path=n/a interface=n/a member=n/a cookie=120 reply_cookie=76 signature=v error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.0 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=77 reply_cookie=0 signature=ss error-name=n>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.0 path=n/a interface=n/a member=n/a cookie=121 reply_cookie=77 signature=v error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.0 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=78 reply_cookie=0 signature=ss error-name=n>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.0 path=n/a interface=n/a member=n/a cookie=122 reply_cookie=78 signature=v error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.0 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=79 reply_cookie=0 signature=ss error-name=n>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.0 path=n/a interface=n/a member=n/a cookie=123 reply_cookie=79 signature=v error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.0 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=80 reply_cookie=0 signature=ss error-name=n>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.0 path=n/a interface=n/a member=n/a cookie=124 reply_cookie=80 signature=v error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.0 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=81 reply_cookie=0 signature=ss error-name=n>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.0 path=n/a interface=n/a member=n/a cookie=125 reply_cookie=81 signature=v error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.0 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=82 reply_cookie=0 signature=ss error-name=n>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.0 path=n/a interface=n/a member=n/a cookie=126 reply_cookie=82 signature=v error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.0 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=83 reply_cookie=0 signature=ss error-name=n>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.0 path=n/a interface=n/a member=n/a cookie=127 reply_cookie=83 signature=v error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.0 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=84 reply_cookie=0 signature=ss error-name=n>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.0 path=n/a interface=n/a member=n/a cookie=128 reply_cookie=84 signature=v error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.0 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=85 reply_cookie=0 signature=ss error-name=n>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.0 path=n/a interface=n/a member=n/a cookie=129 reply_cookie=85 signature=v error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.0 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/user_4065534_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=86 reply_cookie=0 signature=ss error-name=n>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.0 path=n/a interface=n/a member=n/a cookie=130 reply_cookie=86 signature=v error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite audit[303]: USER_AUTH pid=303 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:authentication grantors=pam_rootok acct="nobody" exe="/usr/bin/su" hostname=? addr=? terminal=? res=success'
Dec 23 20:54:31 systemd-testsuite audit[303]: USER_ACCT pid=303 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:accounting grantors=pam_unix acct="nobody" exe="/usr/bin/su" hostname=? addr=? terminal=? res=success'
Dec 23 20:54:31 systemd-testsuite su[303]: (to nobody) root on none
Dec 23 20:54:31 systemd-testsuite audit[302]: USER_ACCT pid=302 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:accounting grantors=pam_unix acct="nobody" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
Dec 23 20:54:31 systemd-testsuite audit[302]: USER_START pid=302 uid=0 auid=65534 ses=1 msg='op=PAM:session_open grantors=pam_loginuid,pam_keyinit,pam_systemd acct="nobody" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
Dec 23 20:54:31 systemd-testsuite systemd[302]: Successfully forked off '(sd-pam)' as PID 304.
Dec 23 20:54:31 systemd-testsuite systemd[1]: Got message type=method_call sender=:1.4 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=GetDynamicUsers cookie=2 reply_cookie=0 signature=n/a error-name=n/a error-messag>
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=method_return sender=n/a destination=:1.4 path=n/a interface=n/a member=n/a cookie=131 reply_cookie=2 signature=a(us) error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite audit[303]: CRED_ACQ pid=303 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:setcred grantors=pam_rootok acct="nobody" exe="/usr/bin/su" hostname=? addr=? terminal=? res=success'
Dec 23 20:54:31 systemd-testsuite audit[303]: USER_START pid=303 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:session_open grantors=pam_unix acct="nobody" exe="/usr/bin/su" hostname=? addr=? terminal=? res=success'
Dec 23 20:54:31 systemd-testsuite su[303]: pam_unix(su:session): session opened for user nobody by (uid=0)
Dec 23 20:54:31 systemd-testsuite systemd[302]: user@65534.service: Executing: /usr/lib/systemd/systemd --user
Dec 23 20:54:31 systemd-testsuite testsuite.sh[305]: Failed to connect to bus: No such file or directory
Dec 23 20:54:31 systemd-testsuite su[303]: pam_unix(su:session): session closed for user nobody
Dec 23 20:54:31 systemd-testsuite audit[303]: USER_END pid=303 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:session_close grantors=pam_unix acct="nobody" exe="/usr/bin/su" hostname=? addr=? terminal=? res=success'
Dec 23 20:54:31 systemd-testsuite audit[303]: CRED_DISP pid=303 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:setcred grantors=pam_rootok acct="nobody" exe="/usr/bin/su" hostname=? addr=? terminal=? res=success'
Dec 23 20:54:31 systemd-testsuite systemd[1]: systemd-journald.service: Received EPOLLHUP on stored fd 52 (stored), closing.
Dec 23 20:54:31 systemd-testsuite systemd[1]: testsuite.service: Control group is empty.
Dec 23 20:54:31 systemd-testsuite systemd[1]: Received SIGCHLD from PID 298 (testsuite.sh).
Dec 23 20:54:31 systemd-testsuite systemd[1]: Child 298 (testsuite.sh) died (code=exited, status=1/FAILURE)
Dec 23 20:54:31 systemd-testsuite systemd[1]: testsuite.service: Child 298 belongs to testsuite.service.
Dec 23 20:54:31 systemd-testsuite systemd[1]: testsuite.service: Main process exited, code=exited, status=1/FAILURE
Dec 23 20:54:31 systemd-testsuite systemd[1]: testsuite.service: Failed with result 'exit-code'.
Dec 23 20:54:31 systemd-testsuite systemd[1]: testsuite.service: Service will not restart (restart setting)
Dec 23 20:54:31 systemd-testsuite systemd[1]: testsuite.service: Changed start -> failed
Dec 23 20:54:31 systemd-testsuite systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=132 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Dec 23 20:54:31 systemd-testsuite systemd[1]: testsuite.service: Job 72 testsuite.service/start finished, result=failed
Dec 23 20:54:31 systemd-testsuite systemd[1]: Failed to start Testsuite service.

```

**Comment by mrc0mmand** [comment]
Hm, it certainly fails consistently, but it also started failing in #11199 :thinking:

**Comment by poettering** [comment]
(Also fixes #14484)

File: src/core/unit.c [hunk 1]
```diff
@@ -3219,11 +3219,10 @@ static int get_name_owner_handler(sd_bus_message *message, void *userdata, sd_bu
         u->get_name_owner_slot = sd_bus_slot_unref(u->get_name_owner_slot);
 
         e = sd_bus_message_get_error(message);
-        if (sd_bus_error_has_name(e, "org.freedesktop.DBus.Error.NameHasNoOwner"))
```
Comment (keszybz) [inline src/core/unit.c:3227]: sd_bus_error_has_name() return false when e==NULL, so this was OK.

**Comment by poettering** [comment]
Hmm, so the later commits then change this around again all the time. If it's OK with you I'll just merge this as is? WE could revert this after, but I am too lazy to do the fiddly rebasing to drop it right-away, and it doesn't reallymatter, does it?

**Comment by poettering** [comment]
@keszybz hmm, i tried to prepare a revert now, to see how this goes, but the code is very different now, I don't think it makes sense to revert it anymore.  hope that's ok?

**Pull request merged**